yorm/utilities.py

```diff
@@ -37,10 +41,81 @@ def find(class_or_instance, *args, create=False, **kwargs):  # pylint: disable=r
         return None
 
 
-def match(cls, **kwargs):
-    """Yield all matching mapped objects."""
-    log.debug((cls, kwargs))
-    raise NotImplementedError
+class GlobFormatter(string.Formatter):
+    """
+    Uses '*' for all unknown fields
+    """
+
+    WILDCARD = object()
+
+    def get_field(self, field_name, args, kwargs):
+        try:
+            return super().get_field(field_name, args, kwargs)
+        except (KeyError, IndexError, AttributeError):
+            return self.WILDCARD, None
+
+    def get_value(self, key, args, kwargs):
+        try:
+            return super().get_value(key, args, kwargs)
+        except (KeyError, IndexError, AttributeError):
+            return self.WILDCARD
+
+    def convert_field(self, value, conversion):
+        if value is self.WILDCARD:
+            return self.WILDCARD
+        else:
+            return super().convert_field(value, conversion)
+
+    def format_field(self, value, format_spec):
+        if value is self.WILDCARD:
+            return '*'
+        else:
+            return super().format_field(value, format_spec)
+
+
+def _unpack_parsed_fields(pathfields):
+    return {
+        (k[len('self.'):] if k.startswith('self.') else k): v
+        for k, v in pathfields.items()
+    }
+
+
+def match(cls_or_path, factory=None, **kwargs):
+    """match(class, [callable], ...) -> instance, ...
+    match(str, callable, ...) -> instance, ...
+
+    Yield all matching mapped objects. Can be used two ways:
+    * With a YORM-decorated class, optionally with a factory callable
+    * With a Python 3-style string template and a factory callable
+
+    The factory callable must accept keyuword arguments, extracted from the file
+    name merged with those passed to match(). If no factory is given, the class
+    itself is used as the factory (same signature).
+    """
+    if isinstance(cls_or_path, type):
+        path_format = common.path_formats[cls_or_path]
+        # Let KeyError fail through
+        if factory is None:
+            factory = cls_or_path
+    else:
+        path_format = cls_or_path
+        if factory is None:
+            raise TypeError("Factory must be given if a string template is used")
+
+    log.debug((path_format, factory, kwargs))
+    gf = GlobFormatter()
+    mock = types.SimpleNamespace(**kwargs)
```

Comment:
It's a fast way to turn a dict into a pile of attributes, so that when we try to format the string, things like `{self.spam}` actually work (if `spam` was passed in).